### ISSUE TITLE
Some issues about the reproducing results

### ISSUE THREAD

**Issue by yux94** [baidu-research/NCRF#12]
When I tried to reproduce your codes, there are less-than-perfect results.
For example, below is the raw test_001 tiff
![tif_raw_convert_img](https://user-images.githubusercontent.com/27813052/43056324-349ee9b6-8e6e-11e8-88a0-24d69112ee33.jpeg)
And after the whole training with the ResNet18-CRF, I got the test result as:
![probmap_convert_img](https://user-images.githubusercontent.com/27813052/43056354-6db146d6-8e6e-11e8-82a8-0aacacd016b4.jpeg)
while the ground-truth mask is something like:(since the camelyon16 organizers didn't provide the test GT with the format of tiff , I transfered the raw test and xml to the mask with the ASAP software manually.)
![npy_mask_convert_img](https://user-images.githubusercontent.com/27813052/43056361-7c3a706a-8e6e-11e8-8b98-0ce3df79eeb8.jpeg)

And I have just followed your test steps and evaluated the average FROC score for the whole test set, and got this:
![froc_npy](https://user-images.githubusercontent.com/27813052/43056405-c186dd84-8e6e-11e8-91cd-bbafd667d02a.png)

However, the result is not at all satisfying. 

And is there any other trick in your preprocess, postprocess, or the training process?

**Comment by yux94** [comment]
Thank you so much for your generous help and suggestions！
After plotting the prob map with `cmap=jet`， I got the probability map of Test_001.tif with my ckpt as this:
![probmap_convert_img_001_plot](https://user-images.githubusercontent.com/27813052/43112468-e364b20e-8f28-11e8-9108-d484d30e7cf9.png)

Maybe I should train again and check my whole process, thank you so much!

**Comment by yil8** [comment]
@yux94 This one does look worse than my result. In addition to try my ckpt, it would be helpful to also plot your training/validation curve that I can also compare with mine.

**Comment by yux94** [comment]
@yil8 Many thanks!

**Comment by yux94** [comment]
When I try to resample the training patches randomly by myself and train the network again, I got the prob map with test_084 like:
![probmap_convert_img_084_plot_resample_jet](https://user-images.githubusercontent.com/27813052/43305992-3c521930-91ac-11e8-8f4a-b977f9803aaf.png)
And below is the prob map with my previous reproduced ckpt:
![probmap_convert_img_084_plot_reproduce_jet](https://user-images.githubusercontent.com/27813052/43306008-50718c52-91ac-11e8-9709-149527c4631c.png)
And this is your result:
![probmap_convert_img_084_plot_rawbaidu_jet](https://user-images.githubusercontent.com/27813052/43306021-59580bde-91ac-11e8-8a82-1bef27ed14f8.png)
That's very confusing since you have said that one user have already achieved good performance. And I am working on retraining the network again. Besides, it would be very nice if you could provide the detailed process of your sampling with hard mining. ([#14](https://github.com/baidu-research/NCRF/issues/14) )

**Comment by yil8** [comment]
@yux94 When I said other users achieved good performance, I mean they used my provided ckpt and achieved 0.8+ FROC score. Your last heatmap plot based on my ckpt also looks good, and I guess if you calculate the FROC score, it probably will be around 0.8 as well. For the training part, again due to determinism of GPU convolution, it's almost impossible to achieve numerical identical results for retraining. But I would still suggest you plot your training curve, so that I can get some rough ideas. I'm currently traveling for business trip, and will try to find some time to implement the hard-negative sampling part once I'm back to US.

**Comment by yux94** [comment]
Thank you so much for your patient and timely reply.
This is my training curve with 20 epoch, should I train with more epoch till the curve is stable? 
![getimage](https://user-images.githubusercontent.com/27813052/43397377-2f0cb004-9437-11e8-83e1-9ee6f41031c5.png)

**Comment by yux94** [comment]
![getimage 1](https://user-images.githubusercontent.com/27813052/43397944-19b81340-9439-11e8-9539-8c9ca9779bef.png)
And this is my resampling training curve with 20 epoch.

**Comment by yil8** [comment]
@yux94 your first curve looks very similar to mine, which converges to ~0.92 valid accuracy. I guess your second curve does not include hard negative examples, thus it converges to higher accuracy. For the curve with hard negative examples, did you train you model using exactly the same config/command I provided in the README?

**Comment by yux94** [comment]
Yes... Pretty sure.  I will check again , many thanks!

**Comment by yil8** [comment]
@yux94 sorry I couldn't help more on the training side. BTW, what's your FROC score for each case?

**Comment by yux94** [comment]
Sorry for bothering you again, we have tried to use the .ckpt you provided within the codebase to generate the probability map, and the final FROC score is not satisfing either.

FP | 0.25 | 0.5 | 1 | 2 | 4 | 8 | Avg
-- | -- | -- | -- | -- | -- | -- | --
NCRF Model | 0.5265 | 0.6106 | 0.6681 | 0.7257 | 0.7743 | 0.8053 | 0.6851

So we checked the probability maps. First, we generate coordinates of the detected tumor region with the nms.py. Next, we pick out normal cases (48 out pf 129) which is the false positive, and draw the histogram.
![getimage 3](https://user-images.githubusercontent.com/27813052/44715386-461a5480-aaea-11e8-970a-b0afca1b525d.png)

According to this histogram, a good FROC score might be achieved only if the threshold is set to ～0.9.

**Comment by yux94** [comment]
Sorry ,  Test_049 and Test_114 are not excluded , thus I got the bad result.

**Comment by yil8** [comment]
@yux94 did you obtain 0.80+ FROC score after excluding Test_049 and Test_114?

**Comment by Hukongtao** [comment]
@yux94 "I transferred the raw tiff test file and xml file to the tiff mask with the ASAP software manually". How did you do that?I need the test GT,too. I just used the ASAP look the tif.But I don't know how to produce the GT.

**Comment by yux94** [comment]
@yil8  yes, I got the 0.80+ FROC score with your provided` .ckpt` after excluding Test_049 and Test_114, but my reproduced result is not satisfying either.

**Comment by yux94** [comment]
@Hukongtao  First you open the `.tif` file with the ASAP. Next load the `.xml` file and save it as `.araw` file. Then open the `.tif and .araw` file and save it(If I remember correctly). Here is my another solution by `cv2.fillPoly`: https://github.com/yux94/Pathology/blob/master/bin/xml2mask_2.py 
Maybe the second method is more convenient.

**Comment by Hukongtao** [comment]
@yux94 嗯嗯，代码可以使用。我把生成的结果转化成黑白图像得到的是图很大，但是前景只有很小的一块诶。和你在上面展示的不一样

**Comment by Hukongtao** [comment]
@yux94 大神方便留个QQ或者微信么，有些问题还是想向您请教。或者您加我QQ：1821141394

**Comment by yux94** [comment]
Excuse me,  but I have one question again. Did you first train the resnet18 and then finetune the model with the crf model?
